substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/linux/LinuxSubstitutions.java

```diff
@@ -66,6 +72,46 @@ public static String mapLibraryName(String libname) {
     }
 }
 
+@Platforms(Platform.LINUX.class)
+@TargetClass(java.lang.Runtime.class)
+final class Target_java_lang_Runtime {
+
+    @Substitute
+    public int availableProcessors() {
+        if (SubstrateOptions.MultiThreaded.getValue()) {
+            LinuxSched.cpu_set_t cpuSet;
+            int size, res;
+            final int pid = Unistd.getpid();
+            // try easy path
+            cpuSet = StackValue.get(LinuxSched.cpu_set_t.class);
+            size = SizeOf.get(LinuxSched.cpu_set_t.class);
+            res = LinuxSched.sched_getaffinity(pid, size, cpuSet);
+            if (res == 0) return LinuxSched.CPU_COUNT_S(size, cpuSet);
+            // try with more CPUs in a loop
+            size = Integer.highestOneBit(size) << 1;
+            while (Integer.numberOfTrailingZeros(size) < 16) { // we have to give up at *some* point
+                assert Integer.bitCount(size) == 1;
+                cpuSet = calloc(size); // to be safe
+                if (cpuSet.isNull()) throw new InternalError("Cannot determine CPU count");
+                try {
+                    res = LinuxSched.sched_getaffinity(pid, size, cpuSet);
+                    if (res == 0) return LinuxSched.CPU_COUNT_S(size, cpuSet);
+                } finally {
+                    free(cpuSet);
+                }
+                size <<= 1;
+            }
+            // give up
+            res = (int) Unistd.sysconf(Unistd._SC_NPROCESSORS_ONLN());
```

Comment:
Maybe break this into 2 methods: one to get the number of processors from `sched_getaffinity` and one to get the number of processors from `Unistd.sysconf` (and then one that gets the number of processors from the cgroup :-), with a method to call them in some appropriate order.